### PR TITLE
server: support dynamic config max-grpc-send-msg-len and raft-msg-max-batch-size

### LINKED ISSUE #12334: online config: support max-grpc-send-msg-len
### Describe the feature you'd like:

support change grpc message size related configs as follows:
```
set config tikv `server.max-grpc-send-msg-len`=33554432;
set config tikv `server.raft-msg-max-batch-size`=1024;
set config tikv `server.grpc-memory-pool-quota` = 104857600;
set config tikv `raftstore.raft-entry-max-size` = "32MiB";
```

### PR DESCRIPTION
**Pull request by glorv**
Signed-off-by: glorv <glorvs@163.com>

### What is changed and how it works?

Issue Number: Close #12334

What's Changed:

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by glorv** [comment]
@5kbpers  @Connor1996  PTAL

**Review by 5kbpers — APPROVED** [review]

**Comment by 5kbpers** [comment]
CI failed

**Comment by glorv** [comment]
> CI failed

All failures are caused by github connection timeout.😂️

File: src/server/raft_client.rs [hunk 1]
```diff
@@ -201,14 +234,10 @@ impl Buffer for BatchMessageBuffer {
...
-        let mut msg_size = msg.start_key.len()
-            + msg.end_key.len()
-            + msg.get_message().context.len()
-            + msg.extra_ctx.len()
-            // index: 3, term: 2, data tag and size: 3, entry tag and size: 3
-            + 11 * msg.get_message().get_entries().len();
-        for entry in msg.get_message().get_entries() {
-            msg_size += entry.data.len();
+        let msg_size = Self::message_size(&msg);
+        // try refrech config before check
```
Comment (Connor1996) [inline src/server/raft_client.rs:238]: refrech -> refresh

File: src/server/server.rs [hunk 2]
```diff
@@ -163,7 +163,7 @@ impl<T: RaftStoreRouter<E::Local> + Unpin, S: StoreAddrResolver + 'static, E: En
 
         let conn_builder = ConnectionBuilder::new(
             env.clone(),
-            Arc::new(cfg.value().clone()),
+            Arc::clone(cfg),
```
Comment (Connor1996) [inline src/server/server.rs:166]: cfg.clone()

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

File: src/server/raft_client.rs (hunk 1 above)
Comment (glorv) [inline src/server/raft_client.rs:238]:
```suggestion
        // try refresh config before check
```

**Comment by glorv** [comment]
@Connor1996 PTAL again, thanks.

File: components/raftstore/src/store/config.rs [hunk 3]
```diff
@@ -461,6 +460,12 @@ impl Config {
             ));
         }
 
+        if self.raft_entry_max_size.0 == 0 || self.raft_entry_max_size.0 > ReadableSize::gb(3).0 {
+            return Err(box_err!(
+                "raft entry max size should be greater than 0 and less than or equal to 3GiB"
```
Comment (BusyJay) [inline components/raftstore/src/store/config.rs:465]: I don't think we support 3GiB. Not even 2.01GiB.

File: src/server/raft_client.rs [hunk 4]
```diff
@@ -201,14 +234,10 @@ impl Buffer for BatchMessageBuffer {
...
-            + msg.end_key.len()
-            + msg.get_message().context.len()
-            + msg.extra_ctx.len()
-            // index: 3, term: 2, data tag and size: 3, entry tag and size: 3
-            + 11 * msg.get_message().get_entries().len();
-        for entry in msg.get_message().get_entries() {
-            msg_size += entry.data.len();
+        let msg_size = Self::message_size(&msg);
+        // try refresh config before check
+        if let Some(new_cfg) = self.cfg_tracker.any_new() {
```
Comment (BusyJay) [inline src/server/raft_client.rs:239]: Have you bench the performance impact?

File: src/server/config.rs [hunk 5]
```diff
@@ -83,7 +83,6 @@ pub struct Config {
     #[online_config(skip)]
     pub status_thread_pool_size: usize,
 
-    #[online_config(skip)]
```
Comment (BusyJay) [inline src/server/config.rs:86]: I don't see the point of changing this value online. It's only for controlling the size of a batch.

File: src/server/raft_client.rs (hunk 4 above)
Comment (glorv) [reply src/server/raft_client.rs:239]: In my benchmark, this is no notable impact after this change. In the most common case, this is only one extra atomic load.

File: src/server/config.rs (hunk 5 above)
Comment (glorv) [reply src/server/config.rs:86]: The use case is from tidb-cloud to minimize the memory usage in a minimal test cluster. This kind of scenario is not common. I think it is harmless to support it.

File: components/raftstore/src/store/config.rs (hunk 3 above)
Comment (glorv) [reply components/raftstore/src/store/config.rs:465]: I set this to the vlaue as `raft-max-size-per-msg`, this is just an intuitive upper bound, do you mean the hard limit should be 2GB? Sorry I don't find the related code.

File: src/server/raft_client.rs (hunk 4 above)
Comment (BusyJay) [reply src/server/raft_client.rs:239]: Actually the atomic load is way heavier than the original that just sums up several numbers. A better place to check for new version is when a message is sent.

File: src/server/config.rs (hunk 5 above)
Comment (BusyJay) [reply src/server/config.rs:86]: The batch is about 10MiB, and it only consumes about 10MiB * grpc_concurrency = 40MiB.

File: src/server/raft_client.rs (hunk 4 above)
Comment (glorv) [reply src/server/raft_client.rs:239]: So do you mean that we move this check into `flush` call as the frequency of flush should be much lower than `push` especially when the lead is heavy.

Comment (BusyJay) [reply src/server/raft_client.rs:239]: Yes.

**Pull request merged**